src/libs/kata-sys-util/src/mount.rs

```diff
@@ -0,0 +1,1087 @@
+//! specified in mountflags. The choice of which operation to perform is determined by testing the
+//! bits set in mountflags, with the tests being conducted in the order listed here:
+//! - Remount an existing mount: mountflags includes MS_REMOUNT.
+//! - Create a bind mount: mountflags includes MS_BIND.
+//! - Change the propagation type of an existing mount: mountflags includes one of MS_SHARED,
+//!   MS_PRIVATE, MS_SLAVE, or MS_UNBINDABLE.
+//! - Move an existing mount to a new location: mountflags includes MS_MOVE.
+//! - Create a new mount: mountflags includes none of the above flags.
+//!
+//! Since Linux 2.6.26, the MS_REMOUNT flag can be used with MS_BIND to modify only the
+//! per-mount-point flags. This is particularly useful for setting or clearing the "read-only"
+//! flag on a mount without changing the underlying filesystem. Specifying mountflags as:
+//!            MS_REMOUNT | MS_BIND | MS_RDONLY
+//! will make access through this mountpoint read-only, without affecting other mounts.
+//!
+//! # Safety
+//!
+//! Mount related operations are sensitive to security flaws, especially when dealing with symlinks.
+//! There are several CVEs related to file path handling, for example
+//! [CVE-2021-30465](https://github.com/opencontainers/runc/security/advisories/GHSA-c3xm-pvg7-gh7r).
+//!
+//! So some design rules are adopted here:
+//! - all mount variants (`bind_remount_read_only()`, `bind_mount()`, `Mounter::mount()`) assume
+//!   that all received paths are safe.
+//! - the caller must ensure safe version of `PathBuf` are passed to mount variants.
+//! - `create_mount_destination()` may be used to generated safe `PathBuf` for mount destinations.
+//! - the `safe_path` crate should be used to generate safe `PathBuf` for general cases.
+
+use std::fmt::Debug;
+use std::fs;
+use std::io::{self, BufRead};
+use std::os::raw::c_char;
+use std::os::unix::ffi::OsStrExt;
+use std::os::unix::fs::{DirBuilderExt, OpenOptionsExt};
+use std::os::unix::io::AsRawFd;
+use std::path::{Path, PathBuf};
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use std::time::Instant;
+
+use lazy_static::lazy_static;
+use nix::mount::{mount, MntFlags, MsFlags};
+use nix::{unistd, NixPath};
+
+use crate::fs::is_symlink;
+use crate::sl;
+
+/// Default permission for directories created for mountpoint.
+const MOUNT_PERM: u32 = 0o755;
+
+const PROC_MOUNTS_FILE: &str = "/proc/mounts";
+const PROC_FIELDS_PER_LINE: usize = 6;
+const PROC_DEVICE_INDEX: usize = 0;
+const PROC_PATH_INDEX: usize = 1;
+const PROC_TYPE_INDEX: usize = 2;
+
+// Sadly nix/libc doesn't have UMOUNT_NOFOLLOW although it's there since Linux 2.6.34
+const UMOUNT_NOFOLLOW: i32 = 0x8;
+
+lazy_static! {
+    static ref MAX_MOUNT_PARAM_SIZE: usize =
+        if let Ok(Some(v)) = unistd::sysconf(unistd::SysconfVar::PAGE_SIZE) {
+            v as usize
+        } else {
+            4096
+        };
```

Comment:
Shouldn't this just panic if we can't determine the page size?

Reply:
fixed